File: integrations/llamaindex/typescript/context-rot-detection.ts
/**
 * Context Rot Detection -- LlamaIndex TypeScript Integration
 *
 * Monitors context quality over long-running sessions, detecting instruction
 * drift, contradictions, and staleness. Produces structured health reports
 * compatible with LlamaIndex's evaluation patterns.
 *
 * Pattern: patterns/evaluation/context-rot-detection.md
 */

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

type HealthLevel = "healthy" | "degraded" | "critical";
type RemediationAction =
  | "none"
  | "reinject_instructions"
  | "compact_history"
  | "full_reset";

interface HealthCheckResult {
  readonly dimension: string;
  readonly score: number;
  readonly details: string;
  readonly failedChecks: readonly string[];
}

interface ContextHealthReport {
  readonly timestamp: string;
  readonly overallScore: number;
  readonly level: HealthLevel;
  readonly checks: readonly HealthCheckResult[];
  readonly recommendedAction: RemediationAction;
  readonly messageCount: number;
  readonly contextTokensEstimate: number;
}

interface ChatMessage {
  readonly role: "system" | "user" | "assistant" | "tool";
  readonly content: string;
}

interface InstructionRule {
  readonly rule: string;
  readonly testFn: (messages: ChatMessage[]) => boolean;
}

// ---------------------------------------------------------------------------
// Report formatter
// ---------------------------------------------------------------------------

function reportToContextBlock(report: ContextHealthReport): string {
  const lines = [
    "## Context Health Report",
    `**Score**: ${Math.round(report.overallScore * 100)}% (${report.level})`,
    `**Messages**: ${report.messageCount}`,
    `**Est. Tokens**: ${report.contextTokensEstimate.toLocaleString()}`,
    `**Action**: ${report.recommendedAction}`,
    "",
  ];
  for (const check of report.checks) {
    const status = check.score >= 0.7 ? "PASS" : "FAIL";
    lines.push(
      `- [${status}] ${check.dimension}: ${Math.round(check.score * 100)}%`
    );
    for (const fc of check.failedChecks) {
      lines.push(`  - ${fc}`);
    }
  }
  return lines.join("\n");
}

// ---------------------------------------------------------------------------
// Evaluators
// ---------------------------------------------------------------------------

class InstructionAdherenceEvaluator {
  private readonly rules: readonly InstructionRule[];

  constructor(rules: InstructionRule[]) {
    this.rules = Object.freeze([...rules]);
  }

  evaluate(messages: ChatMessage[]): HealthCheckResult {
    if (this.rules.length === 0) {
      return {
        dimension: "instruction_adherence",
        score: 1.0,
        details: "No instructions to check.",
        failedChecks: [],
      };
    }

    let passed = 0;
    const failed: string[] = [];

    for (const rule of this.rules) {
      try {
        if (rule.testFn(messages)) {
          passed++;
        } else {
          failed.push(`Rule not followed: ${rule.rule}`);
        }
      } catch (e) {
        failed.push(`Check error for '${rule.rule}': ${e}`);
      }
    }

    return {
      dimension: "instruction_adherence",
      score: passed / this.rules.length,
      details: `${passed}/${this.rules.length} rules followed.`,
      failedChecks: Object.freeze(failed),
    };
  }
}

class ContradictionEvaluator {
  evaluate(messages: ChatMessage[]): HealthCheckResult {
    const assertions: Map<string, Array<{ value: string; index: number }>> =
      new Map();
    const contradictions: string[] = [];

    for (let i = 0; i < messages.length; i++) {
      const content = messages[i].content;
      if (!content) continue;

      for (const line of content.split("\n")) {
        const trimmed = line.trim().toLowerCase();
        if (trimmed.startsWith("use ") && trimmed.split(/\s+/).length <= 4) {
          const key = "tool_preference";
          const existing = assertions.get(key) ?? [];
          for (const prev of existing) {
            if (prev.value !== trimmed) {
              contradictions.push(
                `'${key}' is '${prev.value}' (msg ${prev.index}) vs '${trimmed}' (msg ${i})`
              );
            }
          }
          existing.push({ value: trimmed, index: i });
          assertions.set(key, existing);
        }
      }
    }

    const score = Math.max(0, 1.0 - contradictions.length / 5);
    return {
      dimension: "contradiction_scan",
      score,
      details: `${contradictions.length} contradictions found.`,
      failedChecks: Object.freeze(contradictions.slice(0, 10)),
    };
  }
}

class StalenessEvaluator {
  private readonly maxAge: number;

  constructor(maxAge: number = 100) {
    this.maxAge = maxAge;
  }

  evaluate(messages: ChatMessage[]): HealthCheckResult {
    const total = messages.length;
    if (total <= this.maxAge) {
      return {
        dimension: "staleness",
        score: 1.0,
        details: "Context is within freshness window.",
        failedChecks: [],
      };
    }

    const staleCount = total - this.maxAge;
    const staleRatio = staleCount / total;
    const score = Math.max(0, 1.0 - staleRatio);

    const issues: string[] = [];
    if (staleRatio > 0.5) {
      issues.push(
        `${staleCount} messages (${Math.round(staleRatio * 100)}%) beyond freshness window of ${this.maxAge}.`
      );
    }

    return {
      dimension: "staleness",
      score,
      details: `${staleCount}/${total} messages potentially stale.`,
      failedChecks: Object.freeze(issues),
    };
  }
}

// ---------------------------------------------------------------------------
// ContextRotDetector
// ---------------------------------------------------------------------------

class ContextRotDetector {
  private readonly adherence: InstructionAdherenceEvaluator;
  private readonly contradictions: ContradictionEvaluator;
  private readonly staleness: StalenessEvaluator;
  private readonly checkInterval: number;
  private readonly degradedThreshold: number;
  private readonly criticalThreshold: number;
  private messagesSinceCheck: number = 0;
  private reportHistory: ContextHealthReport[] = [];

  constructor(params: {
    rules?: InstructionRule[];
    checkInterval?: number;
    degradedThreshold?: number;
    criticalThreshold?: number;
    maxStalenessAge?: number;
  }) {
    this.adherence = new InstructionAdherenceEvaluator(params.rules ?? []);
    this.contradictions = new ContradictionEvaluator();
    this.staleness = new StalenessEvaluator(params.maxStalenessAge ?? 100);
    this.checkInterval = params.checkInterval ?? 20;
    this.degradedThreshold = params.degradedThreshold ?? 0.7;
    this.criticalThreshold = params.criticalThreshold ?? 0.4;
  }

  onMessage(messages: ChatMessage[]): ContextHealthReport | null {
    this.messagesSinceCheck++;
    if (this.messagesSinceCheck < this.checkInterval) return null;
    this.messagesSinceCheck = 0;
    return this.runCheck(messages);
  }

  runCheck(messages: ChatMessage[]): ContextHealthReport {
    const recent = messages.length > 50 ? messages.slice(-50) : messages;

    const adherenceResult = this.adherence.evaluate(recent);
    const contradictionResult = this.contradictions.evaluate(messages);
    const stalenessResult = this.staleness.evaluate(messages);

    const checks = [adherenceResult, contradictionResult, stalenessResult];

    const weights: Record<string, number> = {
      instruction_adherence: 0.5,
      contradiction_scan: 0.3,
      staleness: 0.2,
    };

    const totalWeight = checks.reduce(
      (sum, c) => sum + (weights[c.dimension] ?? 0.2),
      0
    );
    const overall =
      checks.reduce(
        (sum, c) => sum + c.score * (weights[c.dimension] ?? 0.2),
        0
      ) / totalWeight;

    let level: HealthLevel;
    let action: RemediationAction;

    if (overall >= this.degradedThreshold) {
      level = "healthy";
      action = "none";
    } else if (overall >= this.criticalThreshold) {
      level = "degraded";
      action = "reinject_instructions";
    } else {
      level = "critical";
      action = "full_reset";
    }

    const tokenEstimate = messages.reduce(
      (sum, m) => sum + Math.floor((m.content?.length ?? 0) / 4),
      0
    );

    const report: ContextHealthReport = {
      timestamp: new Date().toISOString(),
      overallScore: overall,
      level,
      checks: Object.freeze(checks),
      recommendedAction: action,
      messageCount: messages.length,
      contextTokensEstimate: tokenEstimate,
    };

    this.reportHistory = [...this.reportHistory, report];
    return report;
  }

  get healthTrend(): readonly number[] {
    return this.reportHistory.map((r) => r.overallScore);
  }

  get isDegrading(): boolean {
    const scores = this.healthTrend;
    if (scores.length < 3) return false;
    const recent = scores.slice(-3);
    return recent.every((s, i) => i === 0 || s < recent[i - 1]);
  }
}

// ---------------------------------------------------------------------------
// Demo
// ---------------------------------------------------------------------------

function main(): void {
  const checkMarkdown = (messages: ChatMessage[]): boolean => {
    const assistantMsgs = messages.filter((m) => m.role === "assistant");
    if (assistantMsgs.length === 0) return true;
    const last = assistantMsgs[assistantMsgs.length - 1].content;
    return last.includes("#") || last.includes("```") || last.includes("**");
  };

  const checkConciseness = (messages: ChatMessage[]): boolean => {
    const assistantMsgs = messages.filter((m) => m.role === "assistant");
    if (assistantMsgs.length === 0) return true;
    return assistantMsgs[assistantMsgs.length - 1].content.length < 2000;
  };

  const detector = new ContextRotDetector({
    rules: [
      { rule: "Use markdown formatting", testFn: checkMarkdown },
      {
        rule: "Keep responses concise (<2000 chars)",
        testFn: checkConciseness,
      },
    ],
    checkInterval: 5,
  });

  const messages: ChatMessage[] = [
    { role: "system", content: "You are a code reviewer. Use markdown." },
  ];

  for (let i = 0; i < 15; i++) {
    messages.push({ role: "user", content: `Review change #${i + 1}` });

    if (i < 8) {
      messages.push({
        role: "assistant",
        content: `## Review #${i + 1}\n\n**Status**: Approved\n\n\`\`\`python\n# looks good\n\`\`\``,
      });
    } else {
      messages.push({
        role: "assistant",
        content: `Change ${i + 1} looks fine. No issues found.`,
      });
    }

    const report = detector.onMessage(messages);
    if (report) {
      console.log(`--- Health check at turn ${i + 1} ---`);
      console.log(reportToContextBlock(report));
      console.log();
    }
  }

  console.log(
    `Health trend: ${detector.healthTrend.map((s) => `${Math.round(s * 100)}%`)}`
  );
  console.log(`Is degrading: ${detector.isDegrading}`);
}

main();

File: integrations/llamaindex/python/context_rot_detection.py
"""
Context Rot Detection -- LlamaIndex Integration

Implements context health monitoring using LlamaIndex's evaluation framework.
Custom evaluator classes check for instruction drift, contradictions, and
staleness in conversation context, producing structured health reports.

Pattern: https://github.com/context-engineering-handbook/patterns/evaluation/context-rot-detection.md
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Callable

from llama_index.core.llms import ChatMessage, MessageRole


# ---------------------------------------------------------------------------
# Health report types
# ---------------------------------------------------------------------------


class HealthLevel(Enum):
    HEALTHY = "healthy"
    DEGRADED = "degraded"
    CRITICAL = "critical"


class RemediationAction(Enum):
    NONE = "none"
    REINJECT_INSTRUCTIONS = "reinject_instructions"
    COMPACT_HISTORY = "compact_history"
    FULL_RESET = "full_reset"


@dataclass(frozen=True)
class HealthCheckResult:
    """Result of a single health check dimension."""

    dimension: str
    score: float  # 0.0 to 1.0
    details: str
    failed_checks: tuple[str, ...] = ()


@dataclass(frozen=True)
class ContextHealthReport:
    """Aggregate health report across all dimensions."""

    timestamp: str
    overall_score: float
    level: HealthLevel
    checks: tuple[HealthCheckResult, ...]
    recommended_action: RemediationAction
    message_count: int
    context_tokens_estimate: int

    def to_context_block(self) -> str:
        """Format the report for injection into context."""
        lines = [
            "## Context Health Report",
            f"**Score**: {self.overall_score:.0%} ({self.level.value})",
            f"**Messages**: {self.message_count}",
            f"**Est. Tokens**: {self.context_tokens_estimate:,}",
            f"**Action**: {self.recommended_action.value}",
            "",
        ]
        for check in self.checks:
            status = "PASS" if check.score >= 0.7 else "FAIL"
            lines.append(f"- [{status}] {check.dimension}: {check.score:.0%}")
            for fc in check.failed_checks:
                lines.append(f"  - {fc}")
        return "\n".join(lines)


# ---------------------------------------------------------------------------
# Health check evaluators (modeled after LlamaIndex's evaluator pattern)
# ---------------------------------------------------------------------------


class InstructionAdherenceEvaluator:
    """Tests whether the model still follows key instructions.

    Modeled after LlamaIndex's ResponseEvaluator pattern -- each evaluator
    has an `evaluate` method that returns a structured result.

    In production, you could use an LLM to judge adherence:
        from llama_index.core.evaluation import FaithfulnessEvaluator
    """

    def __init__(self, rules: list[dict[str, Any]]) -> None:
        """
        Args:
            rules: List of dicts with 'rule' (str) and 'test_fn'
                   (callable that takes list[ChatMessage] and returns bool).
        """
        self._rules = rules

    def evaluate(self, messages: list[ChatMessage]) -> HealthCheckResult:
        """Check instruction adherence against recent messages."""
        if not self._rules:
            return HealthCheckResult(
                dimension="instruction_adherence",
                score=1.0,
                details="No instructions to check.",
            )

        passed = 0
        failed: list[str] = []

        for rule in self._rules:
            rule_name = rule["rule"]
            test_fn: Callable = rule["test_fn"]
            try:
                if test_fn(messages):
                    passed += 1
                else:
                    failed.append(f"Rule not followed: {rule_name}")
            except Exception as e:
                failed.append(f"Check error for '{rule_name}': {e}")

        score = passed / len(self._rules)
        return HealthCheckResult(
            dimension="instruction_adherence",
            score=score,
            details=f"{passed}/{len(self._rules)} rules followed.",
            failed_checks=tuple(failed),
        )


class ContradictionEvaluator:
    """Scans context for contradictory statements."""

    def evaluate(self, messages: list[ChatMessage]) -> HealthCheckResult:
        assertions: dict[str, list[tuple[str, int]]] = {}
        contradictions: list[str] = []

        for i, msg in enumerate(messages):
            content = msg.content or ""
            for line in content.split("\n"):
                trimmed = line.strip().lower()
                if trimmed.startswith("use ") and len(trimmed.split()) <= 4:
                    key = "tool_preference"
                    if key in assertions:
                        for prev_value, prev_idx in assertions[key]:
                            if prev_value != trimmed:
                                contradictions.append(
                                    f"'{key}' is '{prev_value}' (msg {prev_idx}) "
                                    f"vs '{trimmed}' (msg {i})"
                                )
                        assertions[key].append((trimmed, i))
                    else:
                        assertions[key] = [(trimmed, i)]

        max_contradictions = 5
        score = max(0.0, 1.0 - len(contradictions) / max_contradictions)

        return HealthCheckResult(
            dimension="contradiction_scan",
            score=score,
            details=f"{len(contradictions)} contradictions found.",
            failed_checks=tuple(contradictions[:10]),
        )


class StalenessEvaluator:
    """Checks whether information in context is outdated."""

    def __init__(self, max_age_messages: int = 100) -> None:
        self._max_age = max_age_messages

    def evaluate(self, messages: list[ChatMessage]) -> HealthCheckResult:
        total = len(messages)
        if total <= self._max_age:
            return HealthCheckResult(
                dimension="staleness",
                score=1.0,
                details="Context is within freshness window.",
            )

        stale_count = total - self._max_age
        stale_ratio = stale_count / total
        score = max(0.0, 1.0 - stale_ratio)

        issues: list[str] = []
        if stale_ratio > 0.5:
            issues.append(
                f"{stale_count} messages ({stale_ratio:.0%}) are beyond "
                f"the freshness window of {self._max_age} messages."
            )

        return HealthCheckResult(
            dimension="staleness",
            score=score,
            details=f"{stale_count}/{total} messages potentially stale.",
            failed_checks=tuple(issues),
        )


# ---------------------------------------------------------------------------
# Context Rot Detector (orchestrator)
# ---------------------------------------------------------------------------


class ContextRotDetector:
    """Orchestrates periodic context health checks.

    Integrates with LlamaIndex's chat engine by accepting ChatMessage lists.
    Call `on_message()` after each turn to trigger periodic checks.

    Usage with a LlamaIndex chat engine:
        from llama_index.core.chat_engine import SimpleChatEngine

        detector = ContextRotDetector(
            rules=[
                {"rule": "Use markdown", "test_fn": check_markdown_usage},
                {"rule": "Include citations", "test_fn": check_citations},
            ],
            check_interval=10,
        )

        # In your chat loop:
        response = chat_engine.chat(user_input)
        report = detector.on_message(chat_engine.chat_history)
        if report and report.recommended_action != RemediationAction.NONE:
            # Trigger remediation
            ...
    """

    def __init__(
        self,
        rules: list[dict[str, Any]] | None = None,
        check_interval: int = 20,
        degraded_threshold: float = 0.7,
        critical_threshold: float = 0.4,
        max_staleness_age: int = 100,
    ) -> None:
        self._adherence = InstructionAdherenceEvaluator(rules or [])
        self._contradictions = ContradictionEvaluator()
        self._staleness = StalenessEvaluator(max_staleness_age)
        self._check_interval = check_interval
        self._degraded_threshold = degraded_threshold
        self._critical_threshold = critical_threshold
        self._messages_since_check = 0
        self._history: list[ContextHealthReport] = []

    def on_message(
        self, messages: list[ChatMessage]
    ) -> ContextHealthReport | None:
        """Call after each message. Returns a report if a check is triggered."""
        self._messages_since_check += 1
        if self._messages_since_check < self._check_interval:
            return None
        self._messages_since_check = 0
        return self.run_check(messages)

    def run_check(self, messages: list[ChatMessage]) -> ContextHealthReport:
        """Run all health checks and produce a report."""
        recent = messages[-50:] if len(messages) > 50 else messages

        adherence_result = self._adherence.evaluate(recent)
        contradiction_result = self._contradictions.evaluate(messages)
        staleness_result = self._staleness.evaluate(messages)

        checks = [adherence_result, contradiction_result, staleness_result]

        # Weighted average
        weights = {
            "instruction_adherence": 0.5,
            "contradiction_scan": 0.3,
            "staleness": 0.2,
        }
        total_weight = sum(weights.get(c.dimension, 0.2) for c in checks)
        overall = (
            sum(c.score * weights.get(c.dimension, 0.2) for c in checks)
            / total_weight
        )

        if overall >= self._degraded_threshold:
            level = HealthLevel.HEALTHY
            action = RemediationAction.NONE
        elif overall >= self._critical_threshold:
            level = HealthLevel.DEGRADED
            action = RemediationAction.REINJECT_INSTRUCTIONS
        else:
            level = HealthLevel.CRITICAL
            action = RemediationAction.FULL_RESET

        token_estimate = sum(
            len(m.content or "") // 4 for m in messages
        )

        report = ContextHealthReport(
            timestamp=datetime.now(timezone.utc).isoformat(),
            overall_score=overall,
            level=level,
            checks=tuple(checks),
            recommended_action=action,
            message_count=len(messages),
            context_tokens_estimate=token_estimate,
        )

        self._history = [*self._history, report]
        return report

    @property
    def health_trend(self) -> list[float]:
        return [r.overall_score for r in self._history]

    @property
    def is_degrading(self) -> bool:
        scores = self.health_trend
        if len(scores) < 3:
            return False
        recent = scores[-3:]
        return all(recent[i] < recent[i - 1] for i in range(1, len(recent)))


# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------


def main() -> None:
    """Demonstrate Context Rot Detection with LlamaIndex ChatMessages."""

    # Define instruction rules to monitor
    def check_markdown(messages: list[ChatMessage]) -> bool:
        """Check if assistant responses use markdown formatting."""
        assistant_msgs = [m for m in messages if m.role == MessageRole.ASSISTANT]
        if not assistant_msgs:
            return True
        last_msg = assistant_msgs[-1].content or ""
        return "#" in last_msg or "```" in last_msg or "**" in last_msg

    def check_conciseness(messages: list[ChatMessage]) -> bool:
        """Check if assistant responses are reasonably concise."""
        assistant_msgs = [m for m in messages if m.role == MessageRole.ASSISTANT]
        if not assistant_msgs:
            return True
        last_msg = assistant_msgs[-1].content or ""
        return len(last_msg) < 2000

    rules = [
        {"rule": "Use markdown formatting", "test_fn": check_markdown},
        {"rule": "Keep responses concise (<2000 chars)", "test_fn": check_conciseness},
    ]

    detector = ContextRotDetector(
        rules=rules,
        check_interval=5,  # Low interval for demo
    )

    # Simulate a conversation
    messages: list[ChatMessage] = [
        ChatMessage(role=MessageRole.SYSTEM, content="You are a code reviewer. Use markdown."),
    ]

    # Add some turns
    for i in range(15):
        messages.append(
            ChatMessage(role=MessageRole.USER, content=f"Review change #{i + 1}")
        )
        # Simulate good responses early, degrading later
        if i < 8:
            messages.append(
                ChatMessage(
                    role=MessageRole.ASSISTANT,
                    content=f"## Review #{i + 1}\n\n**Status**: Approved\n\n```python\n# looks good\n```",
                )
            )
        else:
            # Later responses drop markdown (simulating instruction drift)
            messages.append(
                ChatMessage(
                    role=MessageRole.ASSISTANT,
                    content=f"Change {i + 1} looks fine. No issues found.",
                )
            )

        report = detector.on_message(messages)
        if report is not None:
            print(f"--- Health check at turn {i + 1} ---")
            print(report.to_context_block())
            print()

    # Show trend
    print(f"Health trend: {[f'{s:.0%}' for s in detector.health_trend]}")
    print(f"Is degrading: {detector.is_degrading}")


if __name__ == "__main__":
    main()
